core/src/main/scala/org/broadinstitute/dsp/workbench/welder/package.scala

```diff
@@ -82,14 +84,35 @@ package object welder {
   //This string will be hashed when stored in GCS metadata
   def lockedByString(bucketName: GcsBucketName, ownerEmail: WorkbenchEmail): String = bucketName.value + ":" + ownerEmail.value
 
-  def hashString(metadata: String): IO[HashedLockedBy] = IO {
+  def hashString(metadata: String): Either[Throwable, HashedLockedBy] = Either.catchNonFatal {
     HashedLockedBy(String.format("%032x", new BigInteger(1, MessageDigest.getInstance("SHA-256").digest(metadata.getBytes("UTF-8")))))
   }
 
-  type StorageLinksCache = Ref[IO, Map[Path, StorageLink]]
-  type MetadataCache = Ref[IO, Map[Path, AdaptedGcsMetadataCache]]
+  type StorageLinksCache = Ref[IO, Map[RelativePath, StorageLink]]
+  type MetadataCache = Ref[IO, Map[RelativePath, AdaptedGcsMetadataCache]]
 
   val gcpObjectType = "text/plain"
 
+  def cachedResource[A, B: Decoder: Encoder](path: Path, blockingEc: ExecutionContext, toTuple: B => List[(A, B)])(
+    implicit logger: Logger[IO], cs: ContextShift[IO]
+  ): Stream[IO, Ref[IO, Map[A, B]]] =
+    for {
+      cached <- readJsonFileToA[IO, List[B]](path).map(ls => ls.flatMap(b => toTuple(b)).toMap).handleErrorWith { error =>
```

Comment:
Is there a risk of losing data here? If there are key collisions, `listOfTuples.toMap` will throw away keys. Do we want some kind of `groupBy` instead?

Ex:
```
// imagine my dataset is 6 ints
scala> val dataset = List(1,2,3,4,5,6)
dataset: List[Int] = List(1, 2, 3, 4, 5, 6)

// I want to group them by odd or even
scala> def toTuple(i: Int): List[(Int, Int)] = List(i % 2 -> i)
toTuple: (i: Int)List[(Int, Int)]

// I've lost most of my data set
scala> dataset.flatMap(toTuple).toMap
res6: scala.collection.immutable.Map[Int,Int] = Map(1 -> 5, 0 -> 6)
```

Reply:
I think for one particular local file path, we should only have 1 storagelink, and 1 metadata cache. New data should overwrite old record.